File: Algorithms/p337_House_Robber_III.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rob(TreeNode* root) {
        pair<int,int> res = robHelper(root);
        return max(res.first,res.second);
    }
    // first: max when root not robbed, second: max when root robbed
    pair<int,int> robHelper(TreeNode* root){
        if(root){
            pair<int,int> left = root->left ? robHelper(root->left) : make_pair(0,0);
            pair<int,int> right = root->right ? robHelper(root->right) : make_pair(0,0);
            return {max(left.first,left.second)+max(right.first,right.second),root->val+left.first+right.first};
        }else{
            return {0,0};
        }
    }
};

File: Algorithms/p337_House_Robber_III.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rob(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        res = self.robHelper(root)
        return max(res)
    
    def robHelper(self, root):
        """
        :rtype: tuple
            [0]: max when root not robbed
            [1]: max when root robbed
        """
        if root:
            left =  self.robHelper(root.left) if root.left else (0,0)
            right = self.robHelper(root.right) if root.right else (0,0)
            return (max(left[0],left[1])+max(right[0],right[1]),root.val+left[0]+right[0])
        else:
            return (0,0)
